joblib/hashing.py

```diff
@@ -220,19 +220,20 @@ def save(self, obj):
             # The object will be pickled by the pickler hashed at the end.
             obj = (klass, ('HASHED', obj.dtype, obj.shape, obj.strides))
         elif isinstance(obj, self.np.dtype):
-            # Atomic dtype objects are interned by their default constructor:
-            # np.dtype('f8') is np.dtype('f8')
-            # This interning is not maintained by a
-            # pickle.loads + pickle.dumps cycle, because __reduce__
-            # uses copy=True in the dtype constructor. This
-            # non-deterministic behavior causes the internal memoizer
-            # of the hasher to generate different hash values
-            # depending on the history of the dtype object.
-            # To prevent the hash from being sensitive to this, we use
-            # .descr which is a full (and never interned) description of
-            # the array dtype according to the numpy doc.
-            klass = obj.__class__
-            obj = (klass, ('HASHED', obj.descr))
+            # numpy.dtype consistent hashing is tricky to get right. This comes
+            # from the fact that atomic np.dtype objects are interned:
+            # ``np.dtype('f4') is np.dtype('f4')``. The situation is complicated by the
+            # fact that this interning does not resist a simple pickle.load/dump
+            # roundtrip:
+            # ``pickle.loads(pickle.dumps(np.dtype('f4'))) is not np.dtype('f4')
+            # Because pickle relies on memoization during pickling, it is easy to
+            # produce different hashes for seemingly identical objects, such as
+            # ``[np.dtype('f4'), np.dtype('f4')]``
+            # and ``[np.dtype('f4'), pickle.loads(pickle.dumps('f4'))]``.
+            # To prevent memoization from interfering with hashing, we make a deep copy
+            # of the to-be-hashed dtype before hashing it.
+            # obj = deepcopy(obj)
```

Comment:
@ogrisel turns out `deepcopy` does not make my test pass, but `pickle.loads(pickle.dumps(obj))` does. I made some pickle string introspection and it seems related to  `string` interning, something that popped up very early in the `joblib`/ `dtype` history (cant retrieve a source for now),

Reply:
Alright...

Reply:
To go to the bottom of this we should look exactly at what `copy.deepcopy` does.